### PR TITLE
ddtrace/internal: guard the global tracer against concurrent use

### PR DESCRIPTION
**Pull request by gbbr**

### CODE REVIEW

File: ddtrace/internal/globaltracer.go [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+)
+
+func SetGlobalTracer(t ddtrace.Tracer) {
+	mu.Lock()
+	defer mu.Unlock()
+	globalTracer = t
+}
+
+func GetGlobalTracer() ddtrace.Tracer {
+	mu.RLock()
```
Comment (ufoot) [inline ddtrace/internal/globaltracer.go]: Could we only use a read-only lock here ?

File: ddtrace/tracer/tracer_test.go [hunk 2]
```diff
@@ -28,11 +28,42 @@ func (t *tracer) newChildSpan(name string, parent *span) *span {
 	return t.StartSpan(name, ChildOf(parent.Context())).(*span)
 }
 
+// TestTracerFrentic does frentic testing in a scenario where the tracer is started
```
Comment (palazzem) [inline ddtrace/tracer/tracer_test.go]: `frenetic`

File: ddtrace/tracer/tracer_test.go [hunk 3]
```diff
@@ -28,11 +28,42 @@ func (t *tracer) newChildSpan(name string, parent *span) *span {
...
+func TestTracerFrentic(t *testing.T) {
+	var wg sync.WaitGroup
+
+	n := 1 << 17
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < n; i++ {
+			Stop()
```
Comment (palazzem) [inline ddtrace/tracer/tracer_test.go]: should we add a `Start()` here or it's already called somewhere else?

File: ddtrace/tracer/tracer_test.go (hunk 2 above)
Comment (gbbr) [reply ddtrace/tracer/tracer_test.go]: Done

File: ddtrace/tracer/tracer_test.go (hunk 3 above)
Comment (gbbr) [reply ddtrace/tracer/tracer_test.go]: Done. Added in a separate goroutine. Revealed extra problems.

File: ddtrace/tracer/tracer_test.go [hunk 4]
```diff
@@ -32,27 +32,45 @@ func (t *tracer) newChildSpan(name string, parent *span) *span {
...
 	go func() {
 		defer wg.Done()
 		for i := 0; i < n; i++ {
-			span := StartSpan("test.span")
-			go func(span Span) {
-				defer wg.Done()
-				time.Sleep(time.Millisecond)
-				span.Finish()
-			}(span)
+			Start(withTransport(&transport))
```
Comment (palazzem) [inline ddtrace/tracer/tracer_test.go]: 👍

**Comment by gbbr** [comment]
FYI this is a good solution but the race is still possible in very rare
cases. I will address the situation tomorrow. The race should be impossible
:-)

**Pull request merged**